core/src/test/scala/org/apache/spark/scheduler/DAGSchedulerSuite.scala

```diff
@@ -3847,10 +3890,201 @@ class DAGSchedulerSuite extends SparkFunSuite with TempLocalSparkContext with Ti
 
     // Job successful ended.
     assert(results === Map(0 -> 11, 1 -> 12))
+  }
+
```

Comment:
Could you also add a test for the finalize task competition? e.g., a finalize task with dealy = 0 should replace the pending finalize task.

Reply:
This could be difficult to test since for tests we are not waiting for the delay (eg: 10 secs) to expire and also not running merge finalization in a separate thread. All the DAGScheduler events are synchronized so once all the tasks complete for a stage shuffle merge finalization would complete in the tests (but in the real code we would wait for the 10 secs or whatever configured) so there if minimum pushes complete then that gets cancelled and replaced with a new `finalize` with delay = 0.

Reply:
Since we already have a way to not `shuffleMergeFinalize` in the tests, leveraging that I added a test for the above case you mentioned as well.